File: mavkit-python/examples/raw_messages.py
import asyncio
import os

import mavkit


async def main():
    bind_addr = os.environ.get("MAVKIT_EXAMPLE_UDP_BIND", "0.0.0.0:14550")
    message_count = int(os.environ.get("MAVKIT_EXAMPLE_RAW_COUNT", "5"))

    async with await mavkit.Vehicle.connect_udp(bind_addr) as vehicle:
        identity = vehicle.identity()
        raw = vehicle.raw()
        stream = raw.subscribe()

        print(f"connected: sys={identity.system_id} comp={identity.component_id}")
        print(f"raw handle: {raw}")
        print(f"listening for {message_count} raw MAVLink messages...")

        for _ in range(message_count):
            message = await stream.recv()
            print(
                f"message_id={message.message_id} name={message.message_name} "
                f"sys={message.system_id} comp={message.component_id} payload_len={len(message.payload)}"
            )


asyncio.run(main())

File: examples/raw_messages.rs
use mavkit::Vehicle;
use tokio_stream::StreamExt;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let bind_addr =
        std::env::var("MAVKIT_EXAMPLE_UDP_BIND").unwrap_or_else(|_| "0.0.0.0:14550".to_string());
    let message_count: usize = std::env::var("MAVKIT_EXAMPLE_RAW_COUNT")
        .ok()
        .and_then(|v| v.parse().ok())
        .unwrap_or(5);

    let vehicle = Vehicle::connect_udp(&bind_addr).await?;
    let identity = vehicle.identity();
    println!("connected: sys={} comp={}", identity.system_id, identity.component_id);
    println!("listening for {message_count} raw MAVLink messages...");

    let mut stream = vehicle.raw().subscribe();
    for _ in 0..message_count {
        if let Some(message) = stream.next().await {
            println!(
                "message_id={} sys={} comp={} payload_len={}",
                message.message_id, message.system_id, message.component_id, message.payload.len(),
            );
        }
    }

    vehicle.disconnect().await?;
    Ok(())
}
